### DISCUSSION TITLE
How to Capture Rendered Bevy Frames and send to Webrtc

### DISCUSSION THREAD

**Discussion by cs50victor** [bevyengine/bevy#10767] [Q&A]
Hey everyone, 
New Bevy Engine user here : ) !

How do i capture the Bevy's rendered frames? I know there's a bevy_renderer plugin but I can I access frames like this ?

https://github.com/cs50victor/kitt2/blob/main/lkgpt/src/main.rs#L2892 

I'm currently capturing image frames from a custom vulkan renderer ^ .

**Comment by ShadowMitia** [comment]
Depends on what you want to do, but you can take screenshots inside bevy: https://docs.rs/bevy/latest/bevy/render/view/window/screenshot/struct.ScreenshotManager.html.

**Reply by cs50victor** [reply]
I want to capture each frame quickly. I also want to run it headless.

**Reply by cs50victor** [reply]
would using screenhsots need me to use a window? (I'm planing on running bevy on a server)

**Reply by ShadowMitia** [reply]
Mmmh it might need a window looking quickly at the docs (but I haven't tried), but the implementation of it could help you out.

**Reply by cs50victor** [reply]
Thanks, will definitely look into it.

**Comment by paul-hansen** [comment]
You can set the [`target`](https://docs.rs/bevy/0.12.0/bevy/render/camera/struct.Camera.html#structfield.target) field on the camera to be an image and access that image like you do any other bevy image/assets.

There's an example that shows how to use this to render the camera to a cube's texture: https://github.com/bevyengine/bevy/blob/main/examples/3d/render_to_texture.rs#L105

Another useful thing to know is that you can use [`Image::try_into_dynamic`](https://docs.rs/bevy/0.12.0/bevy/render/texture/struct.Image.html#method.try_into_dynamic) to convert Bevy images into the rust [`image`](https://docs.rs/image/0.24.7/image/enum.DynamicImage.html) crate's dynamic image. Useful if whatever you are using for webrtc already has interop with the image crate. It also has methods for saving to files etc.

**Reply by cs50victor** [reply]
can this be done in a without a window ( headless ) ?

**Reply by paul-hansen** [reply]
I'm not sure, there is an open issue for that but it hasn't been updated in awhile: https://github.com/bevyengine/bevy/issues/22

There's also this issue that seems like it maybe addressed it? The last comment might be helpful: https://github.com/bevyengine/bevy/issues/3155#issuecomment-1476773393 it shows how to disable the winit plugin and set a manual update rate (Might need updated to work with 0.12 though).

We might just need someone to try it and report back whatever issues they run into, it seems like Bevy is open to supporting it at least.

**Comment by cs50victor** [comment]
@ShadowMitia @paul-hansen . It seems that the bevy_renderer is tied to the window being available.

Hey @cart , is bevy_renderer tightly coupled with bevy_renderer ? I'm trying to run bevy headlessly on a server and stream the 
rendered frames to the client ( some context : https://github.com/cs50victor/kitt2/blob/main/README.md ). 

Is there a way to walk around this or would using a custom render loop like be a better solution? i.e https://github.com/hakolao/bevy_vulkano

**Reply by ShadowMitia** [reply]
I didn't dive deeply, but I think bevy_renderer and bevy_window are tightly coupled. No way around this.
It might not be too difficult to decouple them?

bevy_vulkano looks like it has a windowless mode, it might be easier to setup for you right now.

**Reply by ShadowMitia** [reply]
Oh, I found this PR, you might be able to adapt it: https://github.com/bevyengine/bevy/pull/5550/files

**Reply by cs50victor** [reply]
dang. bummer. I'll play around with the PR you provided. Thanks.

**Comment by cs50victor** [comment]
Figured it out. Thanks @ShadowMitia @paul-hansen 
Hopefully this helps someone else in future.
Here's a link to a working example : https://github.com/mosure/bevy_gaussian_splatting/blob/main/examples/headless.rs
